packages/inter-protocol/src/stakeFactory/stakeFactoryKit.js

```diff
@@ -370,14 +370,17 @@ const potBehavior = {
     const { zcf } = state;
     const { helper } = facets;
     assert(state.open);
-    return zcf.makeInvitation(helper.adjustBalancesHook, 'AdjustBalances');
+    return zcf.makeInvitation(
+      seat => helper.adjustBalancesHook(seat),
+      'AdjustBalances',
+    );
```

Comment:
I don't get an intuition from this about what doesn't work any more.  The method being invoked here is on `helper`, right?  In the old form, `helper.adjustBalancesHook` is just as visible, and it's just as clear what object the method is being invoked on. `seat` is being passed implicitly, but if any methods are called on it in `adjustBalancesHook`, it's still apparent that the object is available for its methods to be looked up. 

What does the change in autobinding behavior require?

Reply:
In the following explanation, I ignore the multi-facet case, as that complexity is a distraction from the immediate question.

On representative `helper`, the expression `helper.adjustBalancesHook(seat)` calls the `adjustBalancesHook` function that the virtual object system synthesized, that wraps the `adjustBalancesHook` behavior method that the user wrote. The synthesized method is only called with the non-context arguments explicitly, so it needs to recover the per-representative `context` somehow, to pass as the additional first argument to the behavior method.

The expression `helper.adjustBalancesHook` by itself, without the open paren, simply extracts the synthesized function that would have been called as a first class function.

Before this PR, the virtual object system synthesizes as `adjustBalancesHook` function per representative, each time a representative is create or recreated. This per-representative synthesized function itself captures the relevant `context`, which is how it can provide it to the behavior method. Thus, the synthesized first-class `adjustBalancesHook` function that `helper.adjustBalancesHook` evaluates to already has all the per-representative state that it needs, and can be called later as a function.

With this PR, the virtual object system only synthesizes an `adjustBalancesHook` function on a per-kind basis and stores it on a shared per-kind prototype object that representatives inherit from. Because the new synthesized method is not per-representative, it cannot capture the `context` it needs. It instead captures a `contextMap` that maps from a representative of that kind to the needed `context`. But to use that to look up the context, it needs to know what representative it is invoked on. 

The expression `foo.adjustBalancesHook(seat)` passes the value of `foo`, whatever that is, as the `this` binding to the function looked up by `foo.adjustBalancesHook`. The synthesized function, hoping that its `this` binding is a legitimate representative, looks it up in its `contextMap` to recover the needed `context`. If its `this` binding is anything other than a legitimate representative of its brand, the lookup will fail, preventing application to an illegitimate representative. In

```js
const adjustBalancesHook = helper.adjustBalancesHook;
adjustBalancesHook(seat);
```

the synthesized function is being called with function call syntax (without the `.`) rather than method call syntax. In this case, the `this` binding is `undefined`. The problem is that the synthesized method that `helper.adjustBalancesHook` is only per kind, and has lost all the representative specific context that the method call syntax would have provided --- by binding the left "operand" of the `.` to the the `this` of the called method.

There is a way to repair this, but I would guess it is even more expensive that what we're doing now. By creating the `prototype.adjustBalancesHook` property as an accessor property rather than a data property, on the expression `helper.adjustBalancesHook` where `helper` inherits from the `adjustBalancesHook` accessor property from this `prototype`. On the expression `helper.adjustBalancesHook` the getter would see the value of `helper` as the `this` of that call to the getter. We could have getter of this accessor dynamically produce a bound method wrapping the synthesized method, where the bound method captures the `this` when the property itself was looked up. But the getter cannot distinguish whether the value it returns will be immediately called, or whether it will be held onto and called later, so it would need to create this representative-specific bound method ***on each call***, which may be even more expensive that creating synthesized methods per representative.

But we don't know. It may be cheaper because the getter would only created the bound functions that are used, whereas before this PR, we're creating synthesized methods per representative whether they are used or not. If we think we really want to preserve the current autobinding behavior, we could measure this an an alternate technique and see how expensive it really is.

Reply:
You just provided me with a great example of a use of autobinding that gets broken by this PR. At https://github.com/Agoric/agoric-sdk/commit/f1210a980eff36d73f2d18dfada85c6c84339041 you pass `feeMint.getFeeIssuerKit` as an argument to `makeZoeStorageManager`. After I rebased, this caused at least 7 CI breakages like https://github.com/Agoric/agoric-sdk/runs/7864694216?check_suite_focus=true , which I try to correct at https://github.com/Agoric/agoric-sdk/pull/5970/commits/3a3a6a5c5e7631c2dc8c925e8f24a0f3a47d791a , replacing

```js
feeMint.getFeeIssuerKit
```
with
```js
feeMintAccess => feeMint.getFeeIssuerKit(feeMintAccess),
```

We'll see shortly if my change makes CI happier ;)

Reply:
Thanks. This is precisely the explanation I was looking for.

Reply:
Thanks. Resolving.